proxy/src/metrics.rs

```diff
@@ -53,107 +102,44 @@ pub async fn task_main(config: &MetricCollectionConfig) -> anyhow::Result<Infall
-                        );
-                    }
-
-                    current_metrics.push((
-                        Ids {
-                            endpoint_id: endpoint_id.to_string(),
-                            branch_id: branch_id.to_string(),
-                        },
-                        (value, Utc::now()),
-                    ));
-                }
-            }
-        }
-    }
-
-    current_metrics
-}
-
 #[instrument(skip_all)]
 async fn collect_metrics_iteration(
     client: &http::ClientWithMiddleware,
-    cached_metrics: &mut HashMap<Ids, (u64, DateTime<Utc>)>,
     metric_collection_endpoint: &reqwest::Url,
     hostname: &str,
+    prev: DateTime<Utc>,
+    now: DateTime<Utc>,
 ) -> anyhow::Result<()> {
     info!(
         "starting collect_metrics_iteration. metric_collection_endpoint: {}",
         metric_collection_endpoint
     );
 
-    let current_metrics = gather_proxy_io_bytes_per_client();
-
-    let metrics_to_send: Vec<Event<Ids>> = current_metrics
+    let metrics_to_send: Vec<(Ids, u64)> = USAGE_METRICS
+        .endpoints
         .iter()
-        .filter_map(|(curr_key, (curr_val, curr_time))| {
-            let mut start_time = *curr_time;
-            let mut value = *curr_val;
-
-            if let Some((prev_val, prev_time)) = cached_metrics.get(curr_key) {
-                // Only send metrics updates if the metric has increased
-                if curr_val > prev_val {
-                    value = curr_val - prev_val;
-                    start_time = *prev_time;
-                } else {
-                    if curr_val < prev_val {
-                        error!("proxy_io_bytes_per_client metric value decreased from {} to {} for key {:?}",
-                        prev_val, curr_val, curr_key);
-                    }
-                    return None;
-                }
-            };
-
-            Some(Event {
-                kind: EventType::Incremental {
-                    start_time,
-                    stop_time: *curr_time,
-                },
-                metric: PROXY_IO_BYTES_PER_CLIENT,
-                idempotency_key: idempotency_key(hostname),
-                value,
-                extra: Ids {
-                    endpoint_id: curr_key.endpoint_id.clone(),
-                    branch_id: curr_key.branch_id.clone(),
-                },
-            })
+        .filter_map(|counter| {
+            // heuristic to see if the branch is still open
+            // if a clone happens while we are observing, the heurstic will be incorrect.
+            //
+            // Worst case is that we won't report an event for this endpoint.
+            // However, for the strong count to be 1 it must have occured that at one instant
+            // all the endpoints were closed, so missing a report because the endpoints are closed is valid.
+            let is_open = Arc::strong_count(&*counter) > 1;
+            let opened = counter.transmitted.swap(0, Ordering::AcqRel);
```

Comment:
The later `if value == 0 && !is_open && opened == 1` would make sense if this was:

```suggestion
            let opened = counter.opened_connections.swap(0, Ordering::AcqRel);
```

Otherwise I cannot understand how could this work. I'm quite confident this should be `opened_connections` and not about the transmitted bytes.

Reply:
Ah yes, typo. good spot